core/src/unprocessed_packet_batches.rs

```diff
@@ -1,209 +1,390 @@
 use {
-    retain_mut::RetainMut,
+    min_max_heap::MinMaxHeap,
     solana_perf::packet::{limited_deserialize, Packet, PacketBatch},
+    solana_runtime::bank::Bank,
     solana_sdk::{
-        hash::Hash, message::Message, short_vec::decode_shortu16_len, signature::Signature,
-        transaction::VersionedTransaction,
+        hash::Hash,
+        message::{Message, VersionedMessage},
+        short_vec::decode_shortu16_len,
+        signature::Signature,
+        transaction::{Transaction, VersionedTransaction},
     },
     std::{
-        collections::{HashMap, VecDeque},
+        cmp::Ordering,
+        collections::{hash_map::Entry, HashMap},
         mem::size_of,
+        rc::Rc,
+        sync::Arc,
     },
+    thiserror::Error,
 };
 
+#[derive(Debug, Error)]
+pub enum DeserializedPacketError {
+    #[error("ShortVec Failed to Deserialize")]
+    // short_vec::decode_shortu16_len() currently returns () on error
+    ShortVecError(()),
+    #[error("Deserialization Error: {0}")]
+    DeserializationError(#[from] bincode::Error),
+    #[error("overflowed on signature size {0}")]
+    SignatureOverflowed(usize),
+}
+
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct ImmutableDeserializedPacket {
+    original_packet: Packet,
+    versioned_transaction: VersionedTransaction,
+    message_hash: Hash,
+    is_simple_vote: bool,
+    fee_per_cu: u64,
+}
+
+impl ImmutableDeserializedPacket {
+    pub fn original_packet(&self) -> &Packet {
+        &self.original_packet
+    }
+
+    pub fn versioned_transaction(&self) -> &VersionedTransaction {
+        &self.versioned_transaction
+    }
+
+    pub fn sender_stake(&self) -> u64 {
+        self.original_packet.meta.sender_stake
+    }
+
+    pub fn message_hash(&self) -> Hash {
+        self.message_hash
```

Comment:
```suggestion
    pub fn message_hash(&self) -> &Hash {
        &self.message_hash
```

I noticed several places we are incurring this copy multiple times in the same scope or only to take the result's reference

Reply:
fixed